TSPL.docc/LanguageGuide/Generics.md

```diff
@@ -1853,6 +1853,286 @@ protocol ComparableContainer: Container where Item: Comparable { }
+
+Other approaches that also have drawbacks
+include taking the arguments an array or a variadic parameter,
+which requires all arguments to be the same type,
+or using `Any` which erases type information.
+
+<!-- XXX transition -->
+
+Writing this function with a parameter pack
+preserves type information about its arguments,
+and lets you call the function with an arbitrary number of arguments:
+
+```swift
+func double<each T: Numeric>(_ value: repeat each T) -> (repeat each T) {
+    // ...
+}
+```
+
+In the code above, `Element` is a generic type parameter.
+It's marked `each Element`,
+indicating that it's a type-parameter pack.
+In contrast to a generic type parameter,
+which serves as a placeholder for a single type,
+a type-parameter pack is a placeholder for multiple types.
+The ability to have `T` contain a varying number of types
+is what allows this version of `double(_:)`
+to take any number of parameters
+while preserving the type information about each of its parameters.
+
+There isn't any syntax in Swift to write out the list of types,
+but you use `repeat` and `each`
+to mark code that is repeated for each value in that list.
+
+```swift
+func double<each T: Numeric>(_ value: repeat each T) -> (repeat each T) {
+    return (repeat (each value).doubled())
+}
+
+
+extension Numeric {
+    func doubled() -> Self {
+        return 2 * self
+    }
+}
+
+let numbers = [12, 0.5, 8 as Int8]
+let doubledNumbers = double(numbers)
+```
+
+The value of `doubledNumbers` is `(24, 1.0, 16)`,
+and each element in this tuple has the same type
+as the value in `numbers` that it comes from.
+Both `12` and `24` are `Int`,
+`0.5` and `1.0` are `Double`,
+and `8` and `16` are `Int8`.
+
+* * *
+
+XXX OUTLINE:
+
+How do you read a parameter pack at its call site?
+
+- A parameter pack "packs" together types.
+  A pack that's made up of types is called a *type pack*.
+  A pack that's made up of values is called a *value pack*.
+
+- A type pack provides the types for a value pack.
+  The corresponding types and values appear at the same positions
+  in their respective packs.
+
+- When you write code that works on collections, you use iteration.
+  Working with parameter packs is similar ---
+  except each element has a different type,
+  and instead of iteration you use repetition.
+
+How do you create a parameter pack?
+
+- In the generic type parameters,
+  write `each` in front of a generic argument
+  to indicate that this argument creates a type parameter pack.
+
+- In the function's parameters,
+  write `repeat` in front of the type for the parameter
+  that can accept a variable number of arguments,
+  to create an *expansion pattern*.
+  You can also write `repeat` in the function's return type.
+
+- The expansion of a repetition pattern produces a comma-separated list.
+  It can appear in generic argument lists,
+  in tuples types,
+  and in function argument lists.
+
+  ```swift
+  func f<each T>(_ t: repeat each T) -> repeat each T
```

Comment:
This example actually is not valid due to including the repetition pattern in the function return type. A valid variation could be:
```swift
func f<each T>(_ t: repeat each T) -> (repeat each T)
```